### PR TITLE
Introduce flag to conditionally disable BES artifact upload

### PR DESCRIPTION
**Pull request by arlyon**
Issue #8673 describes a pain point with the BES when dealing with large output artifacts from RE. This patch introduces a flag to the bazel CLI that conditionally disables BES upload for targets marked with `no-remote` or `no-cache` (even without builds without the bytes).

When the flag `--bep_disable_artifact_upload` is provided, the BES artifact uploader will respect the tags when determining which outputs should be sent to the BES, and ignore files referenced by any events if the target that triggered that event is tagged with either of the aforementioned tags.

This significantly reduces build times in some cases, but is contingent on being able to mark all the offending targets as no-cache or no-remote. I understand that this may not be ideal for all use cases. Some potential solutions:

1) make the flag enable or disable the functionality for _all_ targets, not just those with specific tags (this is the functionality present in the first commit in this patchset)
2) make the tags that enable this functionality configurable
3) introduce a new tag that controls this functionality

Criticisms or opinions welcome.

### CODE REVIEW

**Comment by buchgr** [comment]
@arlyon thanks for the patch!

I'd like to understand the issue that you are trying to solve better:

- Are you using remote execution or remote caching?
- If remote execution, why are you tagging targets with no-remote instead of using builds without the bytes?
- Bazel currently has a bug where it always tries to upload files to the remote cache even if the cache already contains them. If this bug was fixed would you still need the flag you are proposing?

Thanks!

**Comment by arlyon** [comment]
Hey @buchgr 

1) Both, with the BES enabled, but with build without the bytes disabled.
2) I am not able to elaborate.
3) Without remote execution (but with remote caching) the problem does not occur, likely due to the fact that BES and the Remote Cache share the same `ByteStreamUploader`, which is able to prevent duplicate file hashes from being submitted twice. When remote execution is introduced, this hash set is not consistent between the local bazel client and the remote execution server. I assume that that is the bug you are referring to?

After some further discussion today, we discovered that this inconsistency is essentially what causes the symptoms that this PR fixes. A side effect of that being "bursty" event upload, and excessive CAS usage. Obviously fixing the root cause is better than this patch which merely addresses the symptoms. Is there an open issue for this that I can look at? I would be keen to fix it.

**Comment by arlyon** [comment]
I think it may be better to push only the first commit (which globally disables artifact upload), which will resolve the duplicate upload problem until a proper fix is implemented.

**Comment by buchgr** [comment]
@arlyon https://github.com/bazelbuild/bazel/issues/7630

**Comment by buchgr** [comment]
@arlyon there's also `--bes_upload_mode=nowait_for_upload_complete` that might be of interest to you.

**Comment by jjardon** [comment]
Hi @buchgr , I've just wanted to confirm you are ok with going with the first commit of this PR (which globally disables artifact upload and will solve the initial issue), until a more elaborated fix is implemented

**Comment by buchgr** [comment]
I don't think we can do that unfortunately as this will break BES for a lot of people. I think #7630 isn't that hard to fix but I don't have time for it this week :/.

- Add an abstraction that allows sharing a set of hashes between `AbstractRemoteActionCache` and `BuildEventArtifactUploader`.
- For every action output that `AbstractRemoteActionCache` downloads make an entry of its content hash in the set. In BuildEventArtifactUploader check if the hash of the file is in the set before doing the upload. If it's in the set you know the remote side has the data already.

**Comment by arlyon** [comment]
If we were to get that patch up this week, would it be possible to hit the 1.0 merge window? I am starting work on #7630 now, and aim to be done this week. In the case that it isn't, is it possible to get the first commit merged for that deadline with the guarantee that the full fix is in progress?

The first commit is transparent for anyone who doesn't enable the flag and will not break BES. Those who do enable the flag disable artifact upload (keeping event upload in tact) and will only "break" the BES when they are using remote execution without caching at which point the artifacts are never uploaded to the CAS, and the BES references files that do not exist. It is assumed that those who are using the flag are aware of the consequences.

**Comment by arlyon** [comment]
ByteStreamUploader also has some functionality to detect duplicate uploads; is it maybe an idea to have that also consult the same hash set?

**Comment by buchgr** [comment]
1.0 will be cut on September 2nd. I believe that this should be enough time.

**Comment by arlyon** [comment]
Hey @buchgr after further investigation, while your proposal to fix #7630 does partially solve the problem (by removing duplicate uploads), there is an edge case; since the cached blob set is ephemeral, any files that are pulled from the local cache bypass it because they are never downloaded. So, in situations where all the needed files already exist in the local cache, the BES will upload all the artifacts.

A solution to this is to, on start, hash the files in the local cache and to do a "get missing blobs" to determine which ones the CAS has and pre-populate our hashset from that. Does this seem reasonable?

Other than that, the fix is working (and tested).

**Comment by buchgr** [comment]
> A solution to this is to, on start, do a "get missing blobs" on the files in the local cache to determine which ones the CAS has and pre-populate our hashset from that. Does this seem reasonable?

The "get missing blobs" call is a great idea. I believe this might be a good enough general fix and save us from maintaining the hashset. what do you think?

**Comment by arlyon** [comment]
> The "get missing blobs" call is a great idea. I believe this might be a good enough general fix and save us from maintaining the hashset. what do you think?

The implementation I settled on is a hybrid. It maintains a hashset locally of hashes known to be cached, and for the digests it doesn't have an answer for it consults the CAS.

**Comment by buchgr** [comment]
I'd prefer to only do a simple FindMissingBlobs() call in the BytestreamBuildEventArtifactUploader and leave other parts of the code untouched. My reasoning is that it's 1) simple 2) a strict improvement and good enough as far as we know.

**Comment by arlyon** [comment]
Hey @buchgr thanks for the feedback, I've simplified the patch to do missing blob requests. There's a bit of messing around with futures because of a synchronization point due to the requirement that all the submitted are checked in a single request but it is fairly easy to understand if you are familiar with Futures.

Keen to hear your feedback and get this merged soon.

**Comment by buchgr** [comment]
Thanks. Instead of duplicating the code for `getMissingDigests` could you reuse the one in GrpcRemoteCache?

**Comment by arlyon** [comment]
Hey @buchgr I've incorporated your suggestions and fixed the failing CI test case. Let me know what you think.

Alex

File: src/main/java/com/google/devtools/build/lib/remote/CasDigestLookup.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import io.grpc.Context;
+import io.grpc.StatusRuntimeException;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+public class CasDigestLookup {
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/remote/CasDigestLookup.java]: Why have this class? Why no reuse getMissingDigests from GrpcRemoteCache?

File: src/main/java/com/google/devtools/build/lib/remote/ByteStreamBuildEventArtifactUploader.java [hunk 2]
```diff
@@ -48,74 +54,159 @@
...
-                return Futures.transform(
-                    upload, unused -> new PathDigestPair(file, digest), uploadExecutor);
+
+                return Futures.allAsList(cachedPairFutures);
               },
-              MoreExecutors.directExecutor());
-      uploads.add(digestFuture);
-      }
+              MoreExecutors.directExecutor()
+          );
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/remote/ByteStreamBuildEventArtifactUploader.java]: I think the code would be easier to read if the lambdas passed to `Futures.transformAsync` would be methods.

File: src/main/java/com/google/devtools/build/lib/remote/CasDigestLookup.java (hunk 1 above)
Comment (arlyon) [reply src/main/java/com/google/devtools/build/lib/remote/CasDigestLookup.java]: From my interpretation there is a clear boundary between the GrpcRemoteCache and the ByteStreamBuildEventArtifactUploader and I didn't want to have the latter be aware of the former. Perhaps having the GrpcRemoteCache implement a CasDigestLookup interface would be a better approach, I'll do that.

File: src/main/java/com/google/devtools/build/lib/remote/ByteStreamBuildEventArtifactUploader.java (hunk 2 above)
Comment (arlyon) [reply src/main/java/com/google/devtools/build/lib/remote/ByteStreamBuildEventArtifactUploader.java]: Will do.

**Comment by cphang99** [comment]
Hi @buchgr thanks very much for the reviews. Assuming you're happy with this set of reviews, can we get this in before the 1.0 cut (#8573) on Monday?

File: src/main/java/com/google/devtools/build/lib/buildeventstream/BuildEvent.java [hunk 3]
```diff
@@ -53,10 +53,18 @@
 
     public final Path path;
     public final LocalFileType type;
+    public final boolean shouldUpload; /** whether the local file should be uploaded to the BEP */
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/buildeventstream/BuildEvent.java]: afaik this is unused?

Comment (arlyon) [reply src/main/java/com/google/devtools/build/lib/buildeventstream/BuildEvent.java]: It is used for the bep_disable_all_artifact_upload flag here: https://github.com/arlyon/bazel/blob/eb94fe6caddfe89bcc396be556160f30a1bb618f/src/main/java/com/google/devtools/build/lib/remote/ByteStreamBuildEventArtifactUploader.java#L103

Comment (arlyon) [reply src/main/java/com/google/devtools/build/lib/buildeventstream/BuildEvent.java]: And here is where the flag is detected: https://github.com/arlyon/bazel/blob/0a5961bd0e6594800ad5cd9a26a641472d6466cd/src/main/java/com/google/devtools/build/lib/buildeventservice/BuildEventServiceUploader.java#L178

If the flag is provided, any artifacts referenced by build events are not uploaded. Perhaps a better name would be `shouldUploadArtifacts`

**Comment by dslomov** [comment]
Is this really blocking for 1.0?

**Comment by buchgr** [comment]
Yes it's a big performance issue. Will try to get it merged today.

File: src/main/java/com/google/devtools/build/lib/remote/ByteStreamBuildEventArtifactUploader.java [hunk 4]
```diff
@@ -78,44 +89,111 @@ private static boolean isRemoteFile(Path file) {
...
+          toUpload.put(digest, new LinkedList<>());
+        }
+        toUpload.get(digest).add(path);
+      } else {
+        pathDigestFutures.add(Futures.immediateFuture(pathPair.second));
+      }
+    }
+
+    Context prevContext = ctx.attach();
+    ImmutableSet<Digest> missingDigests = digestLookup.getMissingDigests(toUpload.keySet());
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/remote/ByteStreamBuildEventArtifactUploader.java]: That's a blocking call. This won't fly :\.

**Comment by buchgr** [comment]
@arlyon given that this needs to go in to Bazel *today* I have taken the liberty to create a PR based on your change https://github.com/bazelbuild/bazel/pull/9320 that addresses a few issues:

* The getMissingDigests call must not be blocking
* Moves the digestUtil.compute(...) to the thread pool
* Does not introduce a flag to disable BES upload. I am not a fan of the idea as it complicates things quite a bit and I would like to see how far calling getMissingDigests gets us. We may revisit at a later point.

**Comment by arlyon** [comment]
> @arlyon given that this needs to go in to Bazel _today_ I have taken the liberty to create a PR based on your change #9320 that addresses a few issues:
> 
>     * The getMissingDigests call must not be blocking
> 
>     * Moves the digestUtil.compute(...) to the thread pool
> 
>     * Does not introduce a flag to disable BES upload. I am not a fan of the idea as it complicates things quite a bit and I would like to see how far calling getMissingDigests gets us. We may revisit at a later point.

I had the call to `getMissingDigests` return a Future previously (and can revert it to do that) but changed it to just return the raw set because that is what `GrpcRemoteCache` does. Shame about the flag, but understandable. Glad to see this part get merged nonetheless. Feel free to close as you see fit.

**Comment by buchgr** [comment]
This has been merged as https://github.com/bazelbuild/bazel/commit/120cd54c3cdb6e90470ffd3ff275531f776f54a8

**Pull request closed without merging**